ibis/common/caching.py

```diff
@@ -47,56 +49,53 @@ def __init__(
+        self.generate_name = generate_name
         self.key = key or (lambda x: x)
 
+        self.cache: dict[Any, CacheEntry] = dict()
+
     def get(self, key, default=None):
-        try:
-            return self[key]
-        except KeyError:
-            return default
+        if (entry := self.cache.get(key)) is not None:
+            op = entry.ref()
+            return op if op is not None else default
+        return default
 
     def __getitem__(self, key):
-        result = self.cache[key]
-        self.refs[key] += 1
-        return result
+        op = self.cache[key].ref()
+        if op is None:
+            raise KeyError(key)
+        return op
 
-    def store(self, input) -> None:
+    def store(self, input):
         """Compute and store a reference to `key`."""
         key = self.key(input)
-        name = self.names[key]
+        name = self.generate_name()
         self.populate(name, input)
-        self.cache[key] = self.lookup(name)
-        # nothing outside of this instance has referenced this key yet, so the
-        # refcount is zero
-        #
-        # in theory it's possible to call store -> delitem which would raise an
-        # exception, but in practice this doesn't happen because the only call
-        # to store is immediately followed by a call to getitem.
-        self.refs[key] = 0
-
-    def __delitem__(self, key) -> None:
-        # we need to remove the expression representing the computed physical
-        # table as well as the expression that was used to create that table
-        #
-        # bidict automatically handles this for us; without it we'd have to do
-        # to the bookkeeping ourselves with two dicts
-        if (inv_key := self.cache.inverse.get(key)) is None:
-            raise IbisError(
-                "Key has already been released. Did you call "
-                "`.release()` twice on the same expression?"
-            )
-
-        self.refs[inv_key] -= 1
-        assert self.refs[inv_key] >= 0, f"refcount is negative: {self.refs[inv_key]:d}"
-
-        if not self.refs[inv_key]:
-            del self.cache[inv_key], self.refs[inv_key]
-            self.finalize(key)
+        cached = self.lookup(name)
+        finalizer = finalize(cached, self._release, key)
+
+        self.cache[key] = CacheEntry(name, ref(cached), finalizer)
+
+        return cached
+
+    def release(self, name: str) -> None:
+        # Could be sped up with an inverse dictionary,
+        # but explicit release is discouraged, anyway
+        for key, entry in self.cache.items():
+            if entry.name == name:
+                self._release(key)
+                return
+
+        raise IbisError(
+            "Key has already been released. Did you call "
+            "`.release()` twice on the same expression?"
+        )
+
+    def _release(self, key) -> None:
+        entry = self.cache.pop(key)
+        self.finalize(entry.name)
```

Comment:
This call may error for a few reasons:
- Table doesn't exist (I think we should ignore these always, as noted elsewhere).
- Spurious error
- Connection is in weird state since python is finalizing

Since _most_ backends will cleanup temp tables automatically (and errors on exit can give a bad impression of a tool), I wonder if want to do something like:

```python
try:
    self.finalize(entry.name)
except Exception:
    # silence all errors if system is shutting down
    if not sys.is_finalizing():
        raise
```

This way we can still catch bugs in our code since most release calls won't silence everything, but in the case the system is shutting down and e.g. the network is in a weird state we don't see a mountain of failed-to-release-table errors on exit.

Reply:
Ok, I added that.